Optional notification muting

ekiru opened the pull request:
This PR allows users to choose between the current mute behavior for notifications (which mutes posts in timelines, context around other statuses in the detailed view, and mutes notifications) and the old (pre-1.5?) behavior (which is the same except for allowing notifications from the muted user) for each muted user. It should address #4989. It's just a cherry-pick of the implementation of this feature [over in the glitch-soc fork](https://glitch-soc.github.io/docs/features/optional-notification-muting/) (the link contains a description of the UI and behavior and a few screenshots).

The mute modal, pictured below, has a checkbox added to it (with a default label of "Hide notifications from this user?") which defaults to being checked (to match the current behavior). When viewing a list of accounts (such as for muted users, follower/following lists, etc.), there are also now two buttons/icons for muted users instead of just the one: one to unmute and one to toggle whether notifications from the user are muted.

![The new mute modal](https://glitch-soc.github.io/docs/features/optional-notification-muting/mute-dialog.png)

The API changes are fairly minimal:

1. A new `/api/v1/mutes/details` endpoint that returns an array of objects describing the `Mute`s instead of just an array of accounts; the `/api/v1/mutes` endpoint is left unchanged for backward-compatibility's sake and returns all mutes, regardless of whether they mute notifications or not.
2. The `/api/v1/accounts/:id/mute` endpoint accepts an optional `notifications` parameter, which defaults to true and determines whether the mute will mute notifications or not.
3. When the API returns a `Relationship`, if the target account is muted by the user, it returns a JSON object with a boolean property `notifications` rather than just `true` for the `muting` property.

If this gets merged, I'll make a PR to the docs repo to document these changes.

Gargron commented on app/controllers/api/v1/mutes_controller.rb:
```diff
@@ -8,10 +8,15 @@ class Api::V1::MutesController < Api::BaseController
     render json: @accounts, each_serializer: REST::AccountSerializer
   end
 
+  def details
```
How about using `Api::V2::MutesController` instead? e.g. `GET /api/v2/mutes`?

ekiru replied:
Ah, good idea. I moved it to there in 2d003baddafbad335dc2cbf73c3d8db813d62229 and d24bfaae7f15f7548f949b6abb48c3c31d2c0a86

Gargron commented on app/serializers/rest/mute_serializer.rb:
```diff
@@ -0,0 +1,15 @@
+
+  attributes :id, :account, :target_account, :created_at, :hide_notifications
+
+  def account
```
Kinda of questionable whether this is necessary given that you can only ever look at your own mutes?

ekiru replied:
I removed the account from the serializer.

Gargron commented on app/javascript/mastodon/features/ui/components/mute_modal.js:
```diff
@@ -0,0 +1,103 @@
+          <p>
+            <label htmlFor='mute-modal__hide-notifications-checkbox'>
+              <FormattedMessage id='mute_modal.hide_notifications' defaultMessage='Hide notifications from this user?' />
+              <input id='mute-modal__hide-notifications-checkbox' type='checkbox' checked={notifications} onChange={this.toggleNotifications} />
```
You could use the Toggle component for this

ekiru replied:
Would you prefer that? `<input type=checkbox />` felt more appropriate to me, but I don't really have any explicit reason for that, and I suppose it is a little inconsistent with the rest of the UI.

Gargron replied:
Yes, i would prefer consistency. The line height looks a bit too small anyway and this might fix that.

ekiru commented:
Switched to a Toggle but I need to tweak the styling a bit to get it lined up properly; at present it looks like...

<img width="410" alt="screen shot 2017-09-25 at 23 18 28" src="https://user-images.githubusercontent.com/75553/30842521-5e43eda6-a248-11e7-9a32-0737d6361245.png">

ekiru commented:
Now it's still using Toggle but looks a lot better than the screenshot in my comment before this one:

<img width="515" alt="screen shot 2017-09-25 at 23 57 44" src="https://user-images.githubusercontent.com/75553/30843274-a7e2e98a-a24d-11e7-9ca3-029fd0119e4a.png">

Cassolotl commented:
This is kind of a minor comment, but could there be more space between the two rows of text?

beatrix-bitrot commented:
The question now posed by @Gargron is this: is this the ideal way to structure such an endpoint from an API perspective? Once it's added to v2 other new endpoints will be expected to behave consistently with this one etc, so there is an issue of establishing a good precedent/pattern here.

I have invited reviews from @nightpool @unarist @ykzts so that we may have more perspectives on this issue.

akihikodaki commented on app/javascript/mastodon/components/account.js:
```diff
@@ -70,7 +80,18 @@ export default class Account extends ImmutablePureComponent {
-        buttons = <IconButton active icon='volume-up' title={intl.formatMessage(messages.unmute, { name: account.get('username') })} onClick={this.handleMute} />;
+        let hidingNotificationsButton;
+        if (muting.get('notifications')) {
+          hidingNotificationsButton = <IconButton active icon='bell' title={intl.formatMessage(messages.unmute_notifications, { name: account.get('username')  })} onClick={this.handleUnmuteNotifications} />;
```
> `{ name: account.get('username')  }`

The last two spaces are redundant.

akihikodaki commented on spec/controllers/api/v2/mutes_controller_spec.rb:
```diff
@@ -0,0 +1,33 @@
+    end
+
+    it 'returns one mute' do
+      expect(mutes.size).to be(1)
```
Omit parentheses around `be` arguments; keep the style consistent in your changes.

akihikodaki commented on spec/models/concerns/account_interactions_spec.rb:
```diff
@@ -0,0 +1,40 @@
+describe AccountInteractions do
+  describe 'muting an account' do
+    before do
+      @me = Fabricate(:account, username: 'Me')
```
Use `let` in new specs.

beatrix-bitrot commented:
thanks @akihikodaki for your review

@ekiru please address comments and resolve conflicts~

Gargron commented:
So my main concern with the new APIs is that until now our API has been more or less consistent in which entities it returns. There's the big three: Account, Status, Relationship. And all methods operated on Account models instead of the intermediary specific relationship models (I believe that the primary ID of a block does not matter, for example, what matters is account->target_account) which remained hidden. The v2 API exposes those, though only for Mute models. This means for it to become consistent, we'll need to also expose Block and Follow etc etc models. Which makes me uncomfortable.

There is a small precedent for doing it differently. The verify_credentials API method returns your own Account, but it also contains an additional property of your saved defaults.

Another alternative is modifying the Relationship entity such that instead of boolean `true` values it would return option hashes. For example, where previously it would've been

```js
{ muted: true }
```

now would be 

```js
{ muted: { includes_notifications: false } }
```

This would be a breaking API change too, though I suspect it wouldn't be a big deal because a non-empty hash is a truthy value in many languages.

ekiru commented:
I believe I've made all the changes requested by @akihikodaki. I've also rebased off tootsuite/master and fixed the merge conflicts.

I'll make the API changes once I'm more certain what is desired there.

akihikodaki commented on app/javascript/mastodon/reducers/mutes.js:
```diff
@@ -0,0 +1,29 @@
+      state.setIn(['new', 'notifications'], true);
+    });
+  case MUTES_TOGGLE_HIDE_NOTIFICATIONS:
+    return state.setIn(['new', 'notifications'], !state.getIn(['new', 'notifications']));
```
Please use `update` method

akihikodaki commented on app/models/concerns/account_interactions.rb:
```diff
@@ -70,8 +74,14 @@ def block!(other_account)
+    mute = mute_relationships.create_with(hide_notifications: notifications).find_or_create_by!(target_account: other_account)
+    # When toggling a mute between hiding and allowing notifications, the mute will already exist, so the find_or_create_by! call will return the existing Mute without updating the hide_notifications attribute. Therefore, we check that hide_notifications? is what we want and set it if it isn't.
+    if mute.hide_notifications? != notifications
+      mute.hide_notifications = notifications
```
Use `update!` method.

akihikodaki commented on spec/controllers/api/v2/mutes_controller_spec.rb:
```diff
@@ -0,0 +1,33 @@
+      get :index, params: { limit: 1 }
+    end
+
+    let(:mutes) { JSON.parse(response.body) }
```
Use `body_as_json` method.

akihikodaki commented on db/migrate/20170716191202_add_hide_notifications_to_mute.rb:
```diff
@@ -0,0 +1,5 @@
+class AddHideNotificationsToMute < ActiveRecord::Migration[5.1]
+  def change
+    add_column :mutes, :hide_notifications, :boolean, default: false, null: false
```
You should set `default: true` to let all existing mutes to hide notifications.

Gargron replied:
Actually, this should be caught by strong_migrations.. Does that gem not work? Adding a not-null column is a blocking operation. We recently added some helper methods from gitlab to use when an operation like this must be performed

Gargron commented:
@ekiru Yeah, I would much prefer this approach to exposing standalone mute objects. v2 sounds good. I feel like "mute off" should just be "false", not an object, then if there *is* a mute, that's an object with extra options that can be true or false. This reminds me of some Rails options. What the actual properties are called I don't mind, if it's called hide_notifications in the database I don't see a point in renaming it in the API.

aschmitz commented:
Rebased on master, made all of @akihikodaki's suggested changes, as well as @Gargron's non-blocking column addition (although per https://github.com/tootsuite/mastodon/pull/5088#issuecomment-331930812 it wasn't likely to be a big issue either way, better safe than sorry I suppose).

@Gargron: I'm not 100% sure what you're talking about in terms of "mute off" being false: there are a few cases here:

* POST to `/api/v1/accounts/:account_id/mute`
  * With no body: creates or updates a mute for `account_id`, with `hide_notifications` set to true.
  * With a JSON object of `{"notifications":value}`: creates or updates a mute for `account_id`, with `hide_notifications` set to `value`.
* POST to `/api/v1/accounts/:account_id/unmute`
  * With no body: removes a mute for `account_id`. (This feels non-RESTful to me, but whatever, it's the existing API.)
* GET to `/api/v1/mutes`: Returns an array of Account objects which are muted.
* GET to `/api/v2/mutes`: Returns an array of Mute objects, each containing:
  * `id` (an ID)
  * `target_account` (an Account that's muted)
  * `created_at` (a timestamp)
  * `hide_notifications` (whether or not notifications are hidden)

I believe this appears to reflect what you'd discussed above. If there's something else you're thinking of, let me know. (The other potential reading is that you're recommending a new virtual property being added to Account in `mutes` responses, but that seems suboptimal, as it's effectively modifying the type that's being returned. We can talk about that if necessary, though.)

aschmitz commented:
It has been pointed out that perhaps @Gargron meant that the whole `/api/v2/mutes` should be discarded? That's apparently an acceptable solution to @ekiru, so I can effect that if so desired.

The relationships endpoint has the following behavior for the `muting` value:

* If not muting, the value is `false`.
* If muting and hiding notifications, the value is `{"notifications": true}`
* If muting and *not* hiding notifications, the value is `{"notifications": false}`

Which I believe is probably what @Gargron was referring to above.

nightpool commented:
Hows this going? Would love to get this merged post-2.0. Let me know if I can help in any way
